dbms/src/TiDB/Raft/RaftLogManager.cpp

```diff
@@ -0,0 +1,184 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Logger.h>
+#include <Common/Stopwatch.h>
+#include <Common/TiFlashMetrics.h>
+#include <Interpreters/Context.h>
+#include <Storages/Page/V3/Universal/UniversalPageId.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <Storages/Transaction/KVStore.h>
+#include <Storages/Transaction/TMTContext.h>
+#include <Storages/Transaction/Types.h>
+#include <TiDB/Raft/RaftLogManager.h>
+#include <common/defines.h>
+#include <common/logger_useful.h>
+#include <kvproto/raft_serverpb.pb.h>
+
+namespace DB
+{
+
+bool RaftLogEagerGcTasks::updateHint(
+    RegionID region_id,
+    UInt64 eager_truncated_index,
+    UInt64 applied_index,
+    UInt64 threshold)
+{
+    if (applied_index < eager_truncated_index || applied_index - eager_truncated_index < threshold)
+        return false;
+
+    // Try to register a task for eager remove RaftLog to reduce the memory overhead of UniPS
+    std::unique_lock lock(mtx_tasks);
+    if (auto hint_iter = tasks.find(region_id); hint_iter != tasks.end())
+    {
+        if (applied_index > hint_iter->second.applied_index
+            && applied_index - hint_iter->second.applied_index >= threshold)
+        {
+            // LOG_DEBUG(Logger::get(), "update hint, region_id={} applied_index={} hint_applied_index={} diff={}", region_id, applied_index, hint_iter->second.applied_index, applied_index - hint_iter->second.applied_index);
+            // More updates after last GC task is built, merge it into the current GC task
+            hint_iter->second.eager_truncate_index
+                = std::min(hint_iter->second.eager_truncate_index, eager_truncated_index);
+            hint_iter->second.applied_index = std::max(hint_iter->second.applied_index, applied_index);
+            return true;
+        }
+        // No much updates after last GC task is built, let's just keep the task hint unchanged
+        return false;
+    }
+
+    // New GC task hint is created
+    // LOG_DEBUG(Logger::get(), "create hint, region_id={} first_index={} applied_index={}", region_id, first_index, applied_index);
+    tasks[region_id] = RaftLogEagerGcHint{
+        .eager_truncate_index = eager_truncated_index,
+        .applied_index = applied_index,
+    };
+    return true;
+}
+
+RaftLogEagerGcTasks::Hints RaftLogEagerGcTasks::getAndClearHints()
+{
+    std::unique_lock lock(mtx_tasks);
+    Hints hints;
+    hints.swap(tasks);
+    return hints;
+}
+
+struct RegionGcTask
+{
+    bool skip;
+    UInt64 eager_truncate_index;
+    UInt64 applied_index;
+};
+
+RegionGcTask getRegionTask(
```

Comment:
```suggestion
static RegionGcTask getRegionTask(
```

Reply:
it is a "private" function only used in this file, I'll limit its scope